bashlex/subst.py

```diff
@@ -49,7 +49,10 @@ def _parsedolparen(parserobj, base, sindex):
 
 def _extractcommandsubst(parserobj, string, sindex, sxcommand=False):
     if string[sindex] == '(':
-        raise NotImplementedError('arithmetic expansion')
+        node, si = _parsedolparen(parserobj, string, sindex+1)
+        si += 2
+        return ast.node(kind='arithmeticexpansion', arithmetic=node, pos=(sindex-2, si)), si
+        #raise NotImplementedError('arithmetic expansion')
         #return _extractdelimitedstring(parserobj, string, sindex, '$(', '(', '(', sxcommand=True)
```

Comment:
Are you sure you're parsing it correctly? It looks like the Bash implementation called into a different function.

Reply:
In line 3918 of parse.y the parse_comsub function calls parse_matched_pair() to parse the internals of the command subsitituion. In this implementation, parse_matched_pair() is called _parsedolparen(). The function called when paring arithmetic commands in c is parse_arith_cmd() and is defined on line 4640 of parse.y. That function also calls parse_matched_pair to parse the internals of the arithmetic commands. The corresponding parsing functions for command substitutions and arithmetic commands don't actually exist in the python files. _extractcommandsubst is used in its place.  So I added the contents of parse_arith_cmd() to the corresponding location its used in the python implementation. I could implement a wrapper which calls 'parse_arith_cmd() but incurring the functional overhead for the sake of aesthetics doesn't feel worth it and isn't done with respect to command substitutions and a number of other parsing methods.

Reply:
Also parse_matched_pair is defined on line 3545 of parse.y. The function allows for something to wrap the string in question, without being marked as a reserved word. This implementation doesn't allow that so the changes to sindex and si are to drop the final surrounding parenthesis and allow the parser to work without including ( and ) as reserved words inside the arithmetic command's parts member variable.